### ISSUE TITLE
Support StratifiedKFold cross-validation in CatBoost's cv

### ISSUE THREAD

**Issue by gitmario** [catboost/catboost#701]
I need to perform some stratified k-fold cross validations of CatBoost models. I also need to stratify data samples not only using the target variable, but also some categorical predictors (i.e., if the target is {0, 1} and one predictor is "country", I may want to stratify samples both by target and by country, or maybe only by country). For this purpose, I'm using a combination of sklearn's [StratifiedKFold](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.StratifiedKFold.html) and CatBoost. I was wondering if I can use [CatBoost's cv](https://catboost.ai/docs/concepts/python-reference_cv.html) to get a similar functionality.

**Comment by khrisanfov** [comment]
@gitmario StratifiedKFold do not do what you described.

> This cross-validation object is a variation of KFold that returns stratified folds. The folds are made by preserving the percentage of samples for each class.

**Comment by gitmario** [comment]
@khrisanfov yes but it's super-easy to do what I described with StratifiedKFold. It would be easier to integrate with CatBoost cv if, in example, it accepted indexes of the k-folds for cross-validation as an additional argument.

Example:

```
# df = pandas DataFrame with predictors and response
# x = predictor columns of df
# y = response column of df
# stratified_columns = columns that I want to stratify in k-fold cross validation
# y_stratified = combination of stratified columns (concatenation of strings)

k = 5   # 5-fold cross-validation

stratified_columns = ['country', 'response']

skf = StratifiedKFold(n_splits=k,
                      shuffle=True,
                      random_state=int(time.time()))

y_stratified = df[stratified_columns]\
.apply(lambda z: [str(zi) for zi in z], axis=0)\ # convert values to string
.apply(lambda z: '_'.join(z), axis=1) # concatenate strings
    
list_of_train_index = []
list_of_valid_index = []

for train_index, valid_index in skf.split(np.zeros(len(y_stratified)), # Placeholder (dummy input)
                                          y_stratified): # True input
    
    list_of_train_index.append(train_index)
    list_of_valid_index.append(valid_index)
    
# Use train/valid indexes for cross-validation

model = CatBoostRegressor()
    
for j in range(k):
            
    x_train = x.loc[list_of_train_index[j], :]
    x_valid = x.loc[list_of_valid_index[j], :]
    
    y_true_train = y[list_of_train_index[j]]
    y_true_valid = y[list_of_valid_index[j]]
    
    model.fit(X=x_train, y=y_true_train,
              eval_set=(x_valid, y_true_valid),
              cat_features=cat_features)
    
    print(mean_squared_error(y_true_valid, model.predict(x_valid)))
```

**Comment by Sandy4321** [comment]
an interesting question, by the way, is it possible to somehow average models, when each model is created for each split?
for example 
split#1 -> model#1
split#2 -> model#2
split#3 -> model#3
split#4 -> model#4
split#5 -> model#5
it is clear that all models will be different

is catboost combining all this 5 models to one?
or uses combining predictions from these 5 models
for example if for particular observation vector X 
model#1 predicted yes
model#2 predicted yes
model#3 predicted yes
model#4 predicted no
model#1 predicted no

then prediction will be yes, due to a majority votes?
or combining happen using the probability of reliability for each model
or another way?
Thanks

**Comment by khrisanfov** [comment]
@Sandy4321, CatBoost do not do averaging of several models, after cross-validation you need to train final model on all data. But if you need you can do it programmatically in Python.

**Comment by khrisanfov** [comment]
@Sandy4321 by the way you can open issue about it, the developers can implement it in future ;)

**Comment by randomstate13** [comment]
Just save prediction for each split. Because there are too many methods do averaging, e.g. mean, median, voting, gmean etc.

**Comment by andrey-khropov** [comment]
Custom indices for folds have been implemented in commit https://github.com/catboost/catboost/commit/10eb20d0b94d3820011f57db29a94e2e0ced5b92 and included in release 0.16.

**Comment by andrey-khropov** [comment]
Duplicate of #765
